src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/jvm/old_gen/LevelTwoActionBuilder.java

```diff
@@ -0,0 +1,183 @@
+    }
+  }
+
+  private void addShardRequestCacheAction() {
+    Long capacity = NodeConfigCacheReaderUtil
+        .readCacheMaxSizeInBytes(nodeConfigCache, esNode, ResourceEnum.SHARD_REQUEST_CACHE);
+    if (capacity == null) {
+      return;
+    }
+    ModifyCacheMaxSizeAction action = new ModifyCacheMaxSizeAction(esNode, ResourceEnum.SHARD_REQUEST_CACHE,
+        nodeConfigCache, capacity, false, LEVEL_TWO_CONST.CACHE_ACTION_STEP_COUNT);
+    if (action.isActionable()
+        && action.getDesiredCapacityInPercent() >= LEVEL_TWO_CONST.SHARD_REQUEST_CACHE_LOWER_BOUND) {
+      cacheActionMap.put(ResourceEnum.SHARD_REQUEST_CACHE, action);
+    }
+  }
+
+  private void addWriteQueueAction() {
+    Integer capacity = NodeConfigCacheReaderUtil
+        .readQueueCapacity(nodeConfigCache, esNode, ResourceEnum.WRITE_THREADPOOL);
+    if (capacity == null) {
+      return;
+    }
+    ModifyQueueCapacityAction action = new ModifyQueueCapacityAction(esNode, ResourceEnum.WRITE_THREADPOOL,
+        capacity, false, LEVEL_TWO_CONST.QUEUE_ACTION_STEP_COUNT);
+    if (action.isActionable()) {
+      queueActionMap.put(ResourceEnum.WRITE_THREADPOOL, action);
+    }
+  }
+
+  private void addSearchQueueAction() {
+    Integer capacity = NodeConfigCacheReaderUtil
+        .readQueueCapacity(nodeConfigCache, esNode, ResourceEnum.SEARCH_THREADPOOL);
+    if (capacity == null) {
+      return;
+    }
+    ModifyQueueCapacityAction action = new ModifyQueueCapacityAction(esNode, ResourceEnum.SEARCH_THREADPOOL,
+        capacity, false, LEVEL_TWO_CONST.QUEUE_ACTION_STEP_COUNT);
+    if (action.isActionable()) {
+      queueActionMap.put(ResourceEnum.SEARCH_THREADPOOL, action);
+    }
+  }
+
+
+  private void actionPriorityForCache() {
+    actionFilter.put(ResourceEnum.FIELD_DATA_CACHE, true);
+    actionFilter.put(ResourceEnum.SHARD_REQUEST_CACHE, true);
+  }
+
+  // allocate value to its bucket
+  private int bucketization(int lowerBound, int upperBound, int val, int bucketSize) {
+    double step = (double) (upperBound - lowerBound) / (double) bucketSize;
+    return (int) ((double) val / step);
+  }
+
+  // downsize queue based on priority and current queue size
+  private void actionPriorityForQueue() {
+    Integer writeQueueEWMASize = NodeConfigCacheReaderUtil
+        .readQueueEWMASize(nodeConfigCache, esNode, ResourceEnum.WRITE_THREADPOOL);
+    Integer searchQueueEWMASize = NodeConfigCacheReaderUtil
+        .readQueueEWMASize(nodeConfigCache, esNode, ResourceEnum.SEARCH_THREADPOOL);
+    if (writeQueueEWMASize == null || searchQueueEWMASize == null) {
+      return;
+    }
+    ModifyQueueCapacityAction writeQueueAction = queueActionMap.get(ResourceEnum.WRITE_THREADPOOL);
+    ModifyQueueCapacityAction searchQueueAction = queueActionMap.get(ResourceEnum.SEARCH_THREADPOOL);
+    if (writeQueueAction != null && searchQueueAction != null) {
+      int writeQueueSizeBucket = bucketization(writeQueueAction.getLowerBound(),
```

Comment:
I see what you are doing with `bucketization` here, this is good. We should add a doc string to explain to future readers. 

The important point being that we cannot just look at `{w|s}QueueEWMASize/{w|s}QueueCapacity` because if `QueueCapacity` is already very low, say 200 for write queue, and all of it is constantly full (likely with small queues), we will get %age used as ~100% and keep reducing the same small queue. 

Hence we need to look at the full range of sizes available for each queue (lower bound to upper bound), to get a sense of which queue is using more mem. We also don't simply compare absolute values of `{w|s}QueueEWMASize`s because they inherently have different overheads. Comparing them according to their full range ensures we maintain the same proportion while comparing them.

Reply:
done. Created a ThreadPoolConfig object to store default config settings read from rca.conf. Use the lower - upper bounds defined in rca.conf (or the default values ) to calculate bucket size.